lib/iopromise/data_loader.rb

```diff
@@ -3,63 +3,64 @@
 module IOPromise
   module DataLoader
     module ClassMethods
-      def attr_async_data(attr_name, build_func = nil)
-        @promised_data_keys ||= []
-        @promised_data_keys << attr_name
+      def attr_async(attr_name, build_func = nil)
+        @attr_async_names ||= []
+        @attr_async_names << attr_name
 
         if build_func.nil?
           self.class_eval("def async_#{attr_name};@#{attr_name};end")
         else
           self.define_method("async_#{attr_name}") do
-            @promised_data_memo ||= {}
-            @promised_data_memo[attr_name] ||= self.instance_exec(&build_func)
+            @attr_async_memo ||= {}
+            @attr_async_memo[attr_name] ||= self.instance_exec(&build_func)
           end
         end
 
         self.class_eval("def #{attr_name};async_#{attr_name}.sync;end")
       end
   
-      def promised_data_keys
-        @promised_data_keys ||= []
+      def attr_async_names
+        @attr_async_names ||= []
       end
     end
 
     def self.included(base)
       base.extend(ClassMethods)
     end
 
-    def data_promises
-      self.class.promised_data_keys.flat_map do |k|
+    def async_attributes
```

Comment:
I think it makes more sense to have `async_attributes` be a public method, keeping the way we generate the list internally here as an internal behaviour.

This leaves us with:
 * `attr_async_names` being the names of the attributes
 * `attr_async_promises` being the corresponding promises for those attributes (protected for the recursive preload to use)
 * `async_attributes` being the public way of waiting for everything to load like `async_attributes.then {}`, unless the `#sync` is just used/more appropriate.

Does this seem reasonable?

Reply:
This is great! I especially agree with having `attr_async_promises` be protected instead of public.

Reply:
Same, I think this is a great start. Let's see how it feels when used more widely.